src/snarky-typescript/src/index.ts

```diff
@@ -0,0 +1,517 @@
+  abstract ofFieldElements(
+    fieldElements: V[]
+  ): CircuitData<F, Circuit, V, BVar, Type>;
+
+  abstract equals(this: this, that: this | Type): BVar;
+
+  assertEquals(this: this, y: this): void {
+    this.equals(y).assertTrue();
+  }
+
+  abstract if_(
+    this: this,
+    cond: BVar | boolean,
+    that: this | Type
+  ): CircuitData<F, Circuit, V, BVar, Type>;
+
+  abstract read(this: this): Type;
+}
+
+abstract class Circuit<F extends FieldElement<F>, ConstraintKind, Gadget, Result> {
+  abstract Field: Field<F>;
+  abstract Var: CircuitVarCtor<F, this, CircuitVar<F, this>>;
+  abstract Bool: BoolVarCtor<
+    F,
+    this,
+    CircuitVar<F, this>,
+    BoolVar<F, this, CircuitVar<F, this>>
+  >;
+
+  abstract finish(this: this): Result;
+
+  /* Run some code while proving; skipped during constraint generation. */
+  abstract run(this: this, f: () => void): void;
+
+  abstract addConstraint(kind: ConstraintKind, ...data: any): void;
+
+  abstract runGadget(gadget: Gadget): void;
+
+  abstract constant<Type extends AsFieldElements<F, Type>>(
+    this: this,
+    x: Type
+  ): CircuitData<
+    F,
+    this,
+    CircuitVar<F, this>,
+    BoolVar<F, this, CircuitVar<F, this>>,
+    Type
+  >;
+
+  abstract compute<Type extends AsFieldElements<F, Type>>(
+    this: this,
+    x: Type
+  ): CircuitData<
+    F,
+    this,
+    CircuitVar<F, this>,
+    BoolVar<F, this, CircuitVar<F, this>>,
+    Type
+  >;
+  abstract compute<Type extends AsFieldElements<F, Type>>(
+    this: this,
+    f: () => Type
+  ): CircuitData<
+    F,
+    this,
+    CircuitVar<F, this>,
+    BoolVar<F, this, CircuitVar<F, this>>,
+    Type
+  >;
+
+  abstract if_<Var extends AsFieldElements<CircuitVar<F, this>, Var>>(
+    this: this,
+    cond: BoolVar<F, this, CircuitVar<F, this>> | boolean,
+    then_: Var,
+    else_: Var
+  ): Var;
+  abstract if_<Var extends AsFieldElements<CircuitVar<F, this>, Var>>(
+    this: this,
+    cond: BoolVar<F, this, CircuitVar<F, this>> | boolean,
+    args: {
+      then_: Var;
+      else_: Var;
+    }
+  ): Var;
+
+  abstract equals<Var extends AsFieldElements<CircuitVar<F, this>, Var>>(
+    this: this,
+    x: Var,
+    y: Var
+  ): BoolVar<F, this, CircuitVar<F, this>>;
+
+  abstract assertEqual<Var extends AsFieldElements<CircuitVar<F, this>, Var>>(
+    this: this,
+    x: Var,
+    y: Var
+  ): void;
+}
+
+type StandardConstraints = 'multiply' | 'equal' | 'boolean';
+
+function test<
+  F extends FieldElement<F>,
+  C extends Circuit<F, StandardConstraints, undefined, F[]>
+>(Circuit: C): F[] {
+  /* Field interface test. */
+  const f1 = Circuit.Field.one;
+  const f2 = new Circuit.Field(1);
+  const f3 = new Circuit.Field('1');
+  const f4 = f1.mul(2);
+  const f5 = f4.mul(f2);
+  const f6 = Circuit.Field.mul(f2, f3).add(f5);
+
+  /* Circuit interface test. */
+  const knownPublicInput = new Circuit.Var(f6, {isPublic: true});
+  const unknownPublicInput = new Circuit.Var({isPublic: true});
```

Comment:
I'd recommend using a different constructor and avoiding the optional `isPublic: true` parameter (as you mentioned above)